src/placement.c

```diff
@@ -0,0 +1,552 @@
+ * the maximum index j such that edges[j] <= val. The list of edges must be
+ * sorted in increasing order.
+ *
+ * For a returned index j:
+ *
+ * - The index j == -1 implies that val < edges[0].
+ * - An index 0 <= j < (nedges - 1) implies that edges[j] <= val < edges[j + 1].
+ * - The index j == (nedges - 1) implies that edges[nedges - 1] <= val.
+ */
+static int
+find_interval(int *edges, int nedges, double val)
+{
+	int l = 0;
+	int r = nedges;
+
+	while (l < r) {
+		int m = (l + r) / 2;
+		if (edges[m] > val) {
+			r = m;
+		} else {
+			l = m + 1;
+		}
+	}
+
+	return r - 1;
+}
+
+/*
+ * Construct an overlap bitmap for the irregular grid, computed by
+ * build_grid(), that spans view->output. The overlap bitmap maps
+ * each interval to the number of views on the output (excluding *view)
+ * that overlap that interval.
+ */
+static void
+build_overlap(struct overlap_bitmap *bmp, struct view *view)
+{
+	assert(bmp);
+	assert(view);
+
+	struct server *server = view->server;
+	assert(server);
+
+	if (bmp->nr_rows < 1 || bmp->nr_cols < 1) {
+		return;
+	}
+
+	struct output *output = view->output;
+	if (!output_is_usable(output)) {
+		return;
+	}
+
+	struct view *v;
+	for_each_view(v, &server->views, LAB_VIEW_CRITERIA_CURRENT_WORKSPACE) {
+		if (v == view || v->output != output) {
+			continue;
+		}
+
+		/* Find boundaries of the window */
+		struct border margin = ssd_get_margin(v->ssd);
+		int lx = v->pending.x - margin.left;
+		int ly = v->pending.y - margin.top;
+		int hx = v->pending.x + v->pending.width + margin.right;
+		int hy = v->pending.y + v->pending.height + margin.bottom;
+
+		/*
+		 * Find the first and last row and column intervals spanned by
+		 * this view. We want the left and top edges to fall in a
+		 * half-open interval [low, high) but the right and bottom
+		 * edges to fall in a half-open interval (low, high] to ensure
+		 * that the results do not include intervals adjacent to the
+		 * view. View edges are guaranteed by construction to fall
+		 * exactly on the grid points, so we perturb the left and top
+		 * edges by +0.5 units, and the right and bottom edges by -0.5
+		 * units, to ensure that we are always searching in the
+		 * interior of an interval.
+		 */
+
+		/* First row and column overlapping the view */
+		int fc = find_interval(bmp->cols, bmp->nr_cols, lx + 0.5);
+		int fr = find_interval(bmp->rows, bmp->nr_rows, ly + 0.5);
+
+		/* Clip first row/column to start of usable grid */
+		fc = MAX(fc, 0);
+		fr = MAX(fr, 0);
+
+		/* Last row and column overlapping the view */
+		int lc = find_interval(bmp->cols, bmp->nr_cols, hx - 0.5);
+		int lr = find_interval(bmp->rows, bmp->nr_rows, hy - 0.5);
+
+		/*
+		 * Increment the last indices to convert them to strict upper
+		 * bounds, then clip them to the limits of the usable grid.
+		 */
+		lc = MIN(bmp->nr_cols - 1, lc + 1);
+		lr = MIN(bmp->nr_rows - 1, lr + 1);
+
+		/*
+		 * Every interval in the region [fr, lr) x [fc, lc) is
```

Comment:
Mix of `[` and `)`

Reply:
This is intentional; these intervals are half-open, including `fr` and `fc` but excluding `lr` and `lc` (after they've been incremented int he line above).

Reply:
Sorry. Have caught up now. My brain wasn't tuned to interval notation.